File: matlab/kernDiagGradient.m
function g = kernDiagGradient(kern, x, covDiag)

% KERNDIAGGRADIENT Compute the gradient of the kernel's parameters for the diagonal..

% KERN


fileName = [kern.type 'KernDiagGradient'];
if exist(fileName) == 2
  fhandle = str2func(fileName);
  g = fhandle(kern, x, covDiag);
else
  fhandle = str2func([kern.type 'KernGradient']);
  g = zeros(1, kern.nParams);
  for i = 1:size(x, 1)
    g = g ...
        + fhandle(kern, x(i, :), covDiag(i));
  end
end
% Check if parameters are being optimised in a transformed space.
factors = kernFactors(kern, 'gradfact');
g(factors.index) = g(factors.index).*factors.val;

  

File: R/R/kernDiagGradient.R
kernDiagGradient <- function (kern, x, covDiag) {
  x = as.array(x)
  funcName <- paste(kern$type, 'KernDiagGradient', sep="")

  if (exists(funcName)) {
    fhandle = get(funcName, mode="function")
    g = fhandle(kern, x, covDiag)
  } else {
    fhandle = get(paste(kern$type, 'KernGradient', sep=""))
    g = matrix(0, 1, kern$nParams)
    for (i in 1:tail(dim(x),1)) {
      g = g + fhandle(kern, x[i, ], covDiag[i])
    }
  }

  ## Check if parameters are being optimised in a transformed space.
  factors = .kernFactors(kern, 'gradfact')

  for (i in seq(along=factors))
    g[factors[[i]]$index] <- g[factors[[i]]$index]*factors[[i]]$val

  return (g)
}